Improve Buffer and Slice:

nbougalis opened the pull request:
* Make Buffer constructible from a Slice
* Fix self-move-assignment in Buffer
* Add [c]begin/[c]end

nbougalis commented on src/ripple/basics/Buffer.h:
```diff
@@ -58,33 +93,28 @@ class Buffer
-    Buffer (Slice const& slice)
-        : Buffer(slice.data(), slice.size())
+    /** Assign from slice */
+    Buffer& operator= (Slice s)
```
Question: should we have a check here that detects silly uses like:

    std::string test { "Hello World" };

    Buffer b (test.data(), test.size());
    Slice s = b.slice();
    b = s; // oops?

scottschurr replied:
Great question.  Short answer: yes, I think it's worth a check.  I spent some time thinking about this question.  Consider:
```
std::string test { "Hello World" };

Buffer b (test.data(), test.size());
Slice s (b.data(), b.size());
s += 1;
b = s; // copies from released memory.
```
It looks silly when it's all in line like this, but spread across two or more functions this would be an easy mistake to make.

I think the problem is fixable by noting whether `s.data()` sits within the range `&p_[0]` to ` &p_[size_-1]`.  The complexity doesn't need to leak out of this particular assignment operator.  So, yeah, I think it's worth addressing.  Another viable choice would be to remove this assignment operator and only allow construction (not assignment) from a `Slice`.

vinniefalco replied:
`std::string` should use parenthesis to construct, not braces. Its one of the exceptions (because it has a constructor that takes a `std::initializer_list`). `Slice` on the other hand, should construct using braces.

nbougalis replied:
@vinniefalco: do you think that `Buffer` should have code to protect itself from the kind of scenario described here?

vinniefalco replied:
No. The only way I can see to implement that would be to pay a run-time penalty for detecting undefined behavior.

ximinez replied:
What about an `assert`? No run-time penalty on release builds, but if something does go wonky, we're a lot more likely to catch it in development / unit tests.

nbougalis replied:
I have added an `assert`.

scottschurr commented:
Other than the one question you raised I didn't spot any issues.

Just since no good deed should go unpunished, would it be worth adding unit tests for `Buffer` and `Slice`?  Not insisting, just wondering.

ximinez commented on src/ripple/basics/Buffer.h:
```diff
@@ -34,14 +34,49 @@ namespace ripple {
+    /** Create an uninitialized buffer with the given size. */
+    explicit
+    Buffer (std::size_t size)
+        : p_ (size ? new std::uint8_t[size] : nullptr)
```
General question: why construct `p_` with a `new` instead of `make_unique`? i.e.
```
: p_ (size ? std::make_unique<std::uint8_t[]>(size) : nullptr)
```
You get an extra `unique_ptr` which is then moved, but it does seem more C++14ish.

nbougalis replied:
It's just how it was. I could add the `make_unique` but it doesn't really give us anything in this context.

ximinez commented on src/ripple/basics/Buffer.h:
```diff
@@ -34,14 +34,49 @@ namespace ripple {
+        : Buffer (size)
+    {
+        if (size)
+            std::memcpy(p_.get(), data, size);
```
Nit: `memcpy` does the right thing if `size == 0`, doesn't it, so is it worth it to skip the check?

seelabs replied:
@ximinez According to: http://en.cppreference.com/w/cpp/string/byte/memcpy
`If either dest or src is a null pointer, the behavior is undefined, even if count is zero.` So I think the check is still needed since data could be  a null pointer in that case.

ximinez replied:
@seelabs I forgot about that. Thanks!

ximinez commented:
I'd be happier about this if there were some `Buffer` and `Slice` unit tests, particularly since a lot of the new code is uncovered. Codecov appears to ignore those unused functions instead of red-lining them.
* https://codecov.io/gh/ripple/rippled/src/25097bf0da016a60c07467a77d2326562612ac16/src/ripple/basics/Buffer.h (Copy ctor and assignment, `Slice` ctor and assignment, begin, & end.)

Otherwise, looks good. :+1:

vinniefalco commented:
`begin` and `end` are lame on memory buffers

nbougalis commented:
They make them usable with std algorithms. We can remove them, but the alternative is uglier, imo:

    std::find(b.data(), b.data()+b.size(), ...);

No range-for, etc.

vinniefalco commented:
When would you use `std::find` on a memory buffer?

nbougalis commented:
It was an example. There are other things one may want to do - most of the algorithms in `<algorithm>` are very useful and most require "iterators". Only few have `_n` variants.

Since the start and end pointers are frequently needed, and this adds exactly zero overhead compared to getting a pointer and a size, why not have them?

Range-for iterating a buffer is frequently done. Certainly `PublicKey` and `SecretKey`do it a lot when converting to different formats, encoding, decoding, etc.

willwray commented:
## Suggested discussion topic: Small-Buffer optimization
Curious about Buffer sizes, I ran the unit tests with Buffer constructor dumping the size.
(This does not capture the sizes of default constructed Buffers which are assigned afterwards.)
(I removed test `ripple.protocol.STObject parse json array` which constructs Buffer sizes 0...999.)

The most common size is 33 bytes.
The vast majority of sizes are less than 72 byte with a very few at size 255, 256 or 257.

Is this likely to be true of rippled production runs?
If so then this would suggest a Small Buffer optimization with static storage of 72 bytes
(or maybe a few sizes, 36, 72, 260 and remove the default constructor).

The current `Buffer` is just a sized handle to dynamic allocated storage, good for large buffers with `move`. It also handles varying size simply via assignment or reallocation (I don't see the `alloc` member function used anywhere). 

Small Buffers are cheap to copy and with RVO some copies can be elided. Size can be varied up to a static capacity limit. Growth beyond the static limit is only possible by falling back to allocation, a-la small string optimization, but it looks like this complexity is not necessary as the current design has no high level support for growth. So it looks like a static limit is acceptable in practice.

The pay off, of course, is eliminating a bunch of small allocations.

Small Buffers could be based on (or just alias) [boost::static_vector](http://www.boost.org/doc/libs/1_61_0/doc/html/container/non_standard_containers.html#container.non_standard_containers.static_vector), [Docs](https://svn.boost.org/svn/boost/sandbox/static_vector/doc/html/static_vector/introduction.html) a useful array wrapper.
This gives the illusion of being a vector up to the static capacity, providing most vector API including `begin`,`end` (c.f. the review comments above on using algorithms on Buffers).

The results are extracted and plotted in this notebook:
https://gist.github.com/willwray/f888fdb3ac48b13bd9f7a4a3fda2499d
Here's a plot from the notebook (is this viewable here?)
![bufsizebar](https://cloud.githubusercontent.com/assets/3172718/18524084/fb06f422-7a84-11e6-9bbe-8a21349c7bef.png)

nbougalis commented:
@vinniefalco: I took your suggestion and removed `begin` and `end` from both `Slice` and `Buffer`.
@ximinez: I took your suggestion and added some unit tests for `Buffer`; it should have 100% coverage now.

willwray commented on src/test/basics/Buffer_test.cpp:
```diff
@@ -0,0 +1,240 @@
+
+struct Buffer_test : beast::unit_test::suite
+{
+    bool sane (Buffer const& b)
```
Couldn't resist.
```
bool insane (Buffer const& b) { return !b.size() != !b.data(); }
```

ximinez replied:
```bool sane(Buffer const& b) { return static_cast<bool>(b.size()) == static_cast<bool>(b.data()); }```

:bike: :house:

scottschurr commented on src/test/basics/Buffer_test.cpp:
```diff
@@ -0,0 +1,240 @@
+            BEAST_EXPECT(sane(z));
+            BEAST_EXPECT(z == b3);
+
+            // Assign empty slice to empty slice
```
"Assign empty slice to empty buffer"?  Similar comment fixes through the rest of this section I believe.

scottschurr commented:
Love the buffer tests.  Creative solution with the assert on buffer assignment from slice as well.  Thanks.  You may choose to fix up a few comments in the new tests, but not a killer.  :+1:

ximinez commented on src/test/basics/Buffer_test.cpp:
```diff
@@ -0,0 +1,240 @@
+                Buffer y { std::move(x) };
+                BEAST_EXPECT (x.empty());
+                BEAST_EXPECT (sane(x));
+                BEAST_EXPECT (y == b1);
```
```
BEAST_EXPECT(sane(y));
```

nbougalis replied:
Fixed in 787cd8eda0671c42c2bd48abf22d03075c3c61d7

ximinez commented on src/test/basics/Buffer_test.cpp:
```diff
@@ -0,0 +1,240 @@
+        BEAST_EXPECT (sane (b3));
+        BEAST_EXPECT (!b3.empty());
+        BEAST_EXPECT (b3.size() == sizeof(data));
+        BEAST_EXPECT (std::memcmp (b3.data(), data, b3.size()) == 0);
```
IMNSHO, making these `Buffer const` would make the remaining test cases more solid. I pushed a commit for you if you want to use it. https://github.com/ximinez/rippled/commit/4c5cfb9a443dda5f68e87b162045fb5dff186e42

nbougalis commented:
@ximinez: I've added the extra unit test coverage for `clear` and  and `operator(0)`

ximinez commented:
:+1:

nbougalis commented:
Rebased & squashed.

nbougalis commented:
Merged as 4b0d8b630c71215a8721685c685f76c1124df246
